### ISSUE TITLE
StarCoder BetterTransformer - DeepSpeed Zero 3 + A100

### ISSUE THREAD

**Issue by drorbrillsnps** [huggingface/optimum#1316]
### System Info

```shell
I am using the latest optimum version 1.12.0 and latest transformers version 4.32.0 to fine tune StarCoderBase.
The hardware setup is a single node with 4 A100 GPUs and CUDA 11.8
model.to_bettertransformer() is called before training, and before saving the model, model.reverse_bettertransformer() is called.
After the model is saved, I am seeing that pytorch_model.bin is lighter than it should be and weighs 5.4 MB, and the model is unusable.
Pytorch 2.0 and deepspeed 0.10.1 are installed.

I tried Zero stage 2 and stage 3, results are the same.
I also tried "stage3_gather_16bit_weights_on_model_save": true , but that didn't help either.
Thank you
```

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction (minimal, reproducible, runnable)

DeepSpeed configuration:
{
    "bf16": {
      "enabled": "auto"
    },

    "optimizer": {
      "type": "AdamW",
      "params": {
        "lr": "auto",
        "betas": "auto",
        "eps": "auto",
        "weight_decay": "auto"
      }
    },
    "scheduler": {
      "type": "WarmupLR",
      "params": {
        "warmup_min_lr": "auto",
        "warmup_max_lr": "auto",
        "warmup_num_steps": "auto"
      }
    },
    "zero_optimization": {
      "stage": 3,
      "offload_optimizer": {
        "device": "cpu",
        "pin_memory": false
      },
      "offload_param": {
        "device": "cpu",
        "pin_memory": false
      },
      
      "overlap_comm": true,
      "contiguous_gradients": true,
      "sub_group_size": 1e9,
      "reduce_bucket_size": "auto",
      "stage3_prefetch_bucket_size": "auto",
      "stage3_param_persistence_threshold": "auto",
      "stage3_max_live_parameters": 1e9,
      "stage3_max_reuse_distance": 1e9,
      "stage3_gather_16bit_weights_on_model_save": false
    },
    "gradient_accumulation_steps": "auto",
    "gradient_clipping": "auto",
    "steps_per_print": 2000,
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": "auto",
    "wall_clock_breakdown": false
  }

### Expected behavior

Pytorch model should weigh ~59GB , and model should be usable when loaded later from disk.

**Comment by fxmarty** [comment]
@drorbrillsnps Thank you! I can try and help look into it. Could you provide a complete reproduction?

**Comment by drorbrillsnps** [comment]
Thank you!
I am using a slightly modified https://github.com/huggingface/transformers/blob/main/examples/pytorch/language-modeling/run_clm.py training script, where I have added model.to_bettertransformer() and model.reverse_bettertransformer().
deepspeed run_clm.py\
  --model_name_or_path="bigcode/starcoderbase"\
  --dataset_name="<any text will do for debug purposes>"\
  --preprocessing_num_workers=32\
  --save_steps 2\
  --save_total_limit 1\
  --per_device_train_batch_size 1\
  --per_device_eval_batch_size 1\
  --learning_rate 2.0e-5\
  --block_size 8192\
  --gradient_accumulation_steps 16\
  --do_train\
  --do_eval\
  --bf16\
  --lr_scheduler_type="cosine"\
  --evaluation_strategy="epoch"\
  --deepspeed="./deepspeed_z3_config_bf16.json"\
  --report_to="tensorboard"\
  --overwrite_output_dir

**Comment by fxmarty** [comment]
Perfect thank you, I'll look into it tomorrow.

**Comment by drorbrillsnps** [comment]
One more piece of information:
When using "stage3_gather_16bit_weights_on_model_save": true
The process gets stuck, or maybe is in an infinite loop, when saving a checkpoint.

**Comment by fxmarty** [comment]
@drorbrillsnps Thank you I can reproduce the issue.

**Comment by fxmarty** [comment]
@drorbrillsnps The issue appears to arise only with zero3. Using your config with `"stage": 2`, things are fine. Will look into what is wrong.

**Comment by drorbrillsnps** [comment]
Hello @fxmarty , any update on this issue? :)

Thank you!

**Comment by fxmarty** [comment]
Hi @drorbrillsnps for reference from transformers>=4.36 + torch>=2.1.1, Transformers by default uses PyTorch's `scaled_dot_product_attention` operator for GPT BigCode and we deprecate the usage of BetterTransformer for this architecture.

See details at https://huggingface.co/docs/transformers/perf_infer_gpu_one#flashattention-and-memory-efficient-attention-through-pytorchs-scaleddotproductattention and https://github.com/huggingface/transformers/releases/tag/v4.36.0
